### PR TITLE
Lose Crystals on fall

### PR DESCRIPTION
**Pull request by zahranikbakht**

### CODE REVIEW

File: Assets/Scripts/Items/Crystal.cs [hunk 1]
```diff
@@ -5,19 +5,63 @@
...
     {
         this.crystalManager = FindObjectOfType<CrystalManager>();
-
         base.Awake();
     }
 
+    private new void Update()
+    {
+        if (this.isExploding)
+        {
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Items/Crystal.cs:20]: Extract to separate method, `ExplodingUpdate()`

``` 
if (this.isExploding) {
   ExplodingUpdate()
   return;
}

base.Update()

File: Assets/Scripts/Items/Crystal.cs [hunk 2]
```diff
@@ -5,19 +5,63 @@
...
+            var distanceFromUndergroundPlane = Mathf.Abs(this.transform.position.y - crystalManager.underground.UpperLeftCorner.y);   
+            if (distanceFromUndergroundPlane < this.crystalManager.GetHeightOffset())
+                SetupAfterExplosion();
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Items/Crystal.cs:23]: could be named `FinishExplosion`

File: Assets/LoseCrystals.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    private PlayerScore playerScore;
+    [SerializeField] private float radiusOfForce = 200.0f;
+    [SerializeField] private float radiusOfSpawnCircle = .05f;
+    [SerializeField] private float power = 100.0f;
+    [SerializeField] private GameObject crystalPrefab;
+    [SerializeField] private float crystalheightOffsetFromPlayer = 3f;
+    // Start is called before the first frame update
+    void Start()
+    {
+        this.playerScore = this.GetComponent<PlayerScore>();
```
Comment (danielwiktorczyk) [inline Assets/LoseCrystals.cs:17]: this should be in Awake

File: Assets/LoseCrystals.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        
+    }
+
+    public void LoseCoins()
```
Comment (danielwiktorczyk) [inline Assets/LoseCrystals.cs:21]: LoseCrystals, not coins ;)

File: Assets/LoseCrystals.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        SpawnCrystalsBasedOnScore();
+        ExplodeCoins();
```
Comment (danielwiktorczyk) [inline Assets/LoseCrystals.cs:24]: AddExplosionToCrystals

File: Assets/LoseCrystals.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    // Start is called before the first frame update
+    void Start()
+    {
+        this.playerScore = this.GetComponent<PlayerScore>();
+        
+    }
+
+    public void LoseCoins()
+    {
+        SpawnCrystalsBasedOnScore();
```
Comment (danielwiktorczyk) [inline Assets/LoseCrystals.cs:23]: consider RespawnHeldCrystals()

Comment (danielwiktorczyk) [reply Assets/LoseCrystals.cs:23]: I like the extracted methods!!

File: Assets/LoseCrystals.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        ExplodeCoins();
+        this.playerScore.Substract(this.playerScore.CurrentScore);
+    }
+
+    private void ExplodeCoins()
+    {
+        Vector3 explosionPos = transform.position + Vector3.up* crystalheightOffsetFromPlayer;
```
Comment (danielwiktorczyk) [inline Assets/LoseCrystals.cs:30]: check your formatting, + this. also

File: Assets/LoseCrystals.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        Collider[] colliders = Physics.OverlapSphere(explosionPos, radiusOfForce);
+        foreach (Collider hit in colliders)
+        {
+            Rigidbody rb = hit.GetComponent<Rigidbody>();
+
+            if (rb != null)
```
Comment (danielwiktorczyk) [inline Assets/LoseCrystals.cs:36]: woahhhhh hold up. You're not checking if these are crystals! They can be anything with a rb ;) 

Try 
```
var crystal = hit.GetComponent<Crystal>();
if (crystal is null) return;
```

Also, you're iterating over colliders, so instead of `hit` (which is more like the result of a raycast), try `collider`

File: Assets/LoseCrystals.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+            {
+                rb.useGravity = true;
+                rb.velocity = Vector3.zero;
+                rb.AddExplosionForce(power, explosionPos, radiusOfForce, 0.05f,ForceMode.Impulse);
```
Comment (danielwiktorczyk) [inline Assets/LoseCrystals.cs:40]: what's the 0.05?

File: Assets/LoseCrystals.cs [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+            }
+        }
+    }
+
+    private void SpawnCrystalsBasedOnScore()
+    {
+        for (int i = 0; i < this.playerScore.CurrentScore; i++)
+        {
+            var randomizedPos = UnityEngine.Random.insideUnitCircle * radiusOfSpawnCircle;
+            var crystal = Instantiate(crystalPrefab, transform.position + new Vector3(randomizedPos.x, crystalheightOffsetFromPlayer, randomizedPos.y), Quaternion.identity);
```
Comment (danielwiktorczyk) [inline Assets/LoseCrystals.cs:50]: I like to add the position in a new line. If you like, you can try 

```
var randomizedOffset = UnityEngine.Random.insideUnitCircle * radiusOfSpawnCircle;
randomizedOffset .y = crystalheightOffsetFromPlayer

var crystal = Instantiate(crystalPrefab, this.transform); 
crystal.transform.position += randomizedOffset;
```

ALSO, set the new transform's parent to something. Otherwise, it's gonna clog up the hierarchy. Crystals' parents by default is the CrystalManager

File: Assets/Scripts/Scoring/PlayerScore.cs [hunk 11]
```diff
@@ -34,6 +34,7 @@ public void Substract(int amountToSubstract)
             CurrentScore = 0;
 
         this.teamScore.Substract(amountToSubstract);
+        this.playerScoreUI.UpdateScore(CurrentScore);
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Scoring/PlayerScore.cs:37]: oops, I forgot to add this. Thanks!

File: Assets/Scripts/Items/Crystal.cs [hunk 12]
```diff
@@ -5,19 +5,63 @@
...
+            base.Update();
+        }
+        
+    }
+
+    private void SetupAfterExplosion()
+    {
+        this.crystalManager.OnDroppedCrystal(this);
+        isExploding = false;
+        Vector3 newRestPostion = new Vector3(this.transform.position.x, this.crystalManager.underground.UpperLeftCorner.y + this.crystalManager.GetHeightOffset(), this.transform.position.z);
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Items/Crystal.cs:36]: try 
```
var newRestPostion = this.transform.position 
   + Vector3.up * (this.crystalManager.underground.transform.position.y + this.crystalManager.GetHeightOffset());
```

File: Assets/Scripts/Items/Crystal.cs [hunk 13]
```diff
@@ -5,19 +5,63 @@
...
+        {
+            base.Update();
+        }
+        
+    }
+
+    private void SetupAfterExplosion()
+    {
+        this.crystalManager.OnDroppedCrystal(this);
+        isExploding = false;
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Items/Crystal.cs:35]: this.

File: Assets/Scripts/Items/Crystal.cs [hunk 14]
```diff
@@ -5,19 +5,63 @@
...
+        Vector3 newRestPostion = new Vector3(this.transform.position.x, this.crystalManager.underground.UpperLeftCorner.y + this.crystalManager.GetHeightOffset(), this.transform.position.z);
+        this.UpdateRestPosition(newRestPostion);
+        this.GetComponent<Rigidbody>().useGravity = false;
+        this.GetComponent<Rigidbody>().isKinematic = true;
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Items/Crystal.cs:39]: if you're doing multiple calls to rb, or any component in general, set it into a `var`

**Review by danielwiktorczyk — CHANGES_REQUESTED** [review]
Looks cool! I'm going to do a little more testing on my end, the CrystalManager isn't keeping track of the crystals too well rn

File: Assets/Scripts/Items/Crystal.cs (hunk 12 above)
Comment (zahranikbakht) [reply Assets/Scripts/Items/Crystal.cs:36]: notice I am replacing the y position with a new value, not adding to it.

Comment (danielwiktorczyk) [reply Assets/Scripts/Items/Crystal.cs:36]: oops, yah, my b

**Comment by danielwiktorczyk** [comment]
@zahranikbakht , try this: 

Collect some crystals, 
Fall, 
Collect some crystals underground, 
Respawn

After doing this, the crystal that you accumulated underground are now missing from your score!

File: Assets/Scripts/Items/Collectable.cs [hunk 15]
```diff
@@ -23,7 +23,7 @@ protected void Awake()
 
     public void UpdateRestPosition(Vector3 restPosition) => this.restPosition = restPosition;
 
-    private void Update()
+    protected void Update()
```
Comment (eli-lam) [inline Assets/Scripts/Items/Collectable.cs:26]: Why the change to protected?

File: Assets/Scripts/Items/CrystalManager.cs [hunk 16]
```diff
@@ -11,13 +11,13 @@ public class CrystalManager : MonoBehaviour
...
     [SerializeField] private float undergroundBoundsOffset = 5f;
-    
+
     [SerializeField] private float crystalVerticalOffsetHeight = 2f;
     
     [SerializeField] private float spawningCollisionPaddingRadius = 1f;
 
     private PlaneBounds surface;
-    private PlaneBounds underground;
+    public PlaneBounds underground;
```
Comment (eli-lam) [inline Assets/Scripts/Items/CrystalManager.cs:20]: Add the shorthand getter and setters for extra security `{ get; set; }`

File: Assets/Scripts/Items/Crystal.cs [hunk 17]
```diff
@@ -5,19 +5,69 @@
 public class Crystal : Collectable
 {
     private CrystalManager crystalManager;
+    public bool isExploding;
```
Comment (eli-lam) [inline Assets/Scripts/Items/Crystal.cs:8]: `PascalCase` and ` { get; set; }` unless you mean this to be private

**Review by danielwiktorczyk — APPROVED** [review]
Thanks for the changes, looks great! Let's get this explosive commit in there 🚀

File: Assets/Scripts/Underground.cs [hunk 18]
```diff
@@ -37,6 +37,12 @@ void Update()
 
     void Die()
     {
+        loseCrystals.LoseCrystal();
```
Comment (eli-lam) [inline Assets/Scripts/Underground.cs:40]: `this.`

File: Assets/Scripts/Underground.cs [hunk 19]
```diff
@@ -37,6 +37,12 @@ void Update()
...
         this.transform.position = respawnPoint.position;
     }
+    private void OnTriggerEnter(Collider other)
+    {
+        if (other.CompareTag("Underground"))
+            loseCrystals.LoseCrystal();
```
Comment (eli-lam) [inline Assets/Scripts/Underground.cs:46]: `this.`

**Review by eli-lam — CHANGES_REQUESTED** [review]
Just quickly go through your code and prefix with `this.` where you missed them

File: Assets/Scripts/Items/Collectable.cs (hunk 15 above)
Comment (zahranikbakht) [reply Assets/Scripts/Items/Collectable.cs:26]: I'm calling it in the child's Update method

**Review by eli-lam — APPROVED** [review]
